typegate/engine/src/runtimes/wit_wire.rs

```diff
@@ -0,0 +1,316 @@
+// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
+// SPDX-License-Identifier: Elastic-2.0
+
+use crate::interlude::*;
+use dashmap::DashMap;
+#[rustfmt::skip]
+use deno_core as deno_core; // necessary for re-exported macros to work
+use deno_core::OpState;
+
+mod wit {
+    wasmtime::component::bindgen!({
+        path: "../../libs/pyrt_component/wit",
+        async: true,
+    });
+}
+use wasmtime::component::{Component, Linker};
+use wit::exports::metatype::pyrt::mat_wire::{InitArgs, InitError, MatInfo, Req as MatReq};
+use wit::metatype::pyrt::typegate_wire::{Host, Req as HostReq, Res as HostRes};
+
+#[derive(Clone)]
+pub struct Ctx {
+    engine: wasmtime::Engine,
+    instances: Arc<DashMap<String, Instance>>,
+    components: Arc<DashMap<PathBuf, LinkedComponent>>,
+}
+
+#[derive(Clone)]
+struct LinkedComponent(Component, Arc<Linker<TypegateHost>>);
+
+/* static WASI_CTX: once_cell::sync::Lazy<wasmtime_wasi::WasiCtx> = once_cell::sync::Lazy::new(|| {
+    wasmtime_wasi::WasiCtxBuilder::new()
+        .allow_ip_name_lookup(true)
+        .build()
+}); */
+
+impl Ctx {
+    pub fn new(engine: wasmtime::Engine) -> Self {
+        Self {
+            instances: Default::default(),
+            components: Default::default(),
+            engine,
+        }
+    }
+
+    async fn get_component(&self, wasm_relative_path: PathBuf) -> Result<LinkedComponent, String> {
+        let wasm_absolute_path = match std::env::current_dir() {
+            Ok(cwd) => cwd.join(wasm_relative_path),
+            Err(err) => return Err(format!("error trying to find cwd: {err}")),
+        };
+
+        let comp = if let Some(comp) = self.components.get(&wasm_absolute_path) {
+            comp.clone()
+        } else {
+            let engine = self.engine.clone();
+            let path_clone = wasm_absolute_path.clone();
+            let comp = if wasm_absolute_path
+                .extension()
+                .map(|ext| ext == "cwasm")
+                .unwrap_or_default()
+            {
+                // TODO: self manage precompilation cache
```

Comment:
The componentize-py artifact doesn't do the fancy [`wasi-virt` tricks](https://github.com/bytecodealliance/componentize-py/issues/23) or dynamically loading [libpyt](https://github.com/bytecodealliance/componentize-py/issues/28) to minimize the module size. This leads to a lot of work when we compile the component for the first time. Until `componentize-py` sports those features, we can improve startup by using precompilation.

To dedup work, we can have the typegate mantain a local pre-compile cache in `tmp` which should mean that users will only ever need to pre-compile the pyrt module once during dev. 

We can also distribute the `pyrt` pre-compiled for each platform. Or alternatively, use `build.rs` to embed the `pyrt` component into the typegate after pre-compiling and compressing it.

Any thoughts on this?

Reply:
I would go for the later, which I believe makes anyway sense if we keep the wasm rt optional so they can be pulled from a oci / warg repository when needed? Is that much work?

Reply:
The latter as in embed the pre-compiled runtime?

Reply:
yep, at the end of the day, people could also use the runtime as component elsewhere